homeassistant/components/guardian/__init__.py

```diff
@@ -0,0 +1,364 @@
+        """Get updated data from the device."""
+        async with self.client:
+            tasks = [
+                self._async_get_data_from_api(api_category)
+                for api_category in self._api_coros
+            ]
+
+            await asyncio.gather(*tasks)
+
+        LOGGER.debug("Received new data: %s", self.data)
+        async_dispatcher_send(self._hass, TOPIC_UPDATE.format(self.uid))
+
+
+class GuardianEntity(Entity):
+    """Define a base Guardian entity."""
+
+    def __init__(
+        self, guardian: Guardian, kind: str, name: str, device_class: str, icon: str
+    ):
+        """Initialize."""
+        self._attrs = {ATTR_ATTRIBUTION: "Data provided by Elexa"}
+        self._available = True
+        self._device_class = device_class
+        self._guardian = guardian
+        self._icon = icon
+        self._kind = kind
+        self._name = name
+
+    @property
+    def available(self):
+        """Return whether the entity is available."""
+        return bool(self._guardian.data[DATA_PING])
+
+    @property
+    def device_class(self):
+        """Return the device class."""
+        return self._device_class
+
+    @property
+    def device_info(self):
+        """Return device registry information for this entity."""
+        return {
+            "identifiers": {(DOMAIN, self._guardian.uid)},
+            "manufacturer": "Elexa",
+            "model": self._guardian.data[DATA_DIAGNOSTICS]["firmware"],
+            "name": f"Guardian {self._guardian.uid}",
+        }
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self._attrs
+
+    @property
+    def icon(self) -> str:
+        """Return the icon."""
+        return self._icon
+
+    @property
+    def name(self):
+        """Return the name of the entity."""
+        return f"Guardian {self._guardian.uid}: {self._name}"
+
+    @property
+    def should_poll(self) -> bool:
+        """Return True if entity has to be polled for state."""
+        return False
+
+    @property
+    def unique_id(self):
+        """Return the unique ID of the entity."""
+        return f"{self._guardian.uid}_{self._kind}"
+
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+
+        @callback
+        def update():
+            """Update the state."""
+            self.update_from_latest_data()
+            self.async_write_ha_state()
+
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, TOPIC_UPDATE.format(self._guardian.uid), update
+            )
+        )
+
+        await self._guardian.async_register_api_interest(self._kind)
+
+        self.update_from_latest_data()
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Disconnect dispatcher listener when removed."""
+        self._guardian.async_deregister_api_interest(self._kind)
+
+    @callback
+    def update_from_latest_data(self):
```

Comment:
```suggestion
    def _update_from_latest_data(self):
```

Optional: I would have probably made this a protected method since it would never expect to be called outside of the object

Reply:
Sure thing! 👍